Bundled electron application does not start

The application is using published npm packages. It starts in debug mode from VSCode, also starting as an individual node process work. Fails when starting as an application.

I did some "debugging" and it turned out, that the connection is `undefined` when we are creating the proxy workers. Both the message reader and writer looks OK.

```
reader WorkerMessageReader {
  errorEmitter: 
   Emitter {
     _options: undefined,
     _event: [Function],
     _callbacks: CallbackList { _callbacks: [Object], _contexts: [Object] } },
  closeEmitter: 
   Emitter {
     _options: undefined,
     _event: [Function],
     _callbacks: CallbackList { _callbacks: [Object], _contexts: [Object] } },
  partialMessageEmitter: Emitter { _options: undefined },
  worker: 
   Worker {
     domain: null,
     _events: { message: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     exitedAfterDisconnect: undefined,
     suicide: [Getter/Setter],
     state: 'none',
     id: 1,
     process: 
      ChildProcess {
        domain: null,
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        _closesNeeded: 2,
        _closesGot: 0,
        connected: true,
        signalCode: null,
        exitCode: null,
        killed: false,
        spawnfile: '/Users/akos.kitta/Desktop/electron/theia-electron-app-example/dist/mac/theia-electron-app-example.app/Contents/Frameworks/theia-electron-app-example Helper.app/Contents/MacOS/theia-electron-app-example Helper',
        _handle: [Object],
        spawnargs: [Object],
        pid: 13433,
        stdin: null,
        stdout: null,
        stderr: null,
        stdio: [Object],
        channel: [Object],
        _channel: [Getter/Setter],
        _handleQueue: null,
        _pendingHandle: null,
        send: [Function],
        _send: [Function],
        disconnect: [Function],
        _disconnect: [Function] } } }


writer WorkerMessageWriter {
  errorEmitter: 
   Emitter {
     _options: undefined,
     _event: [Function],
     _callbacks: CallbackList { _callbacks: [Object], _contexts: [Object] } },
  closeEmitter: 
   Emitter {
     _options: undefined,
     _event: [Function],
     _callbacks: CallbackList { _callbacks: [Object], _contexts: [Object] } },
  worker: 
   Worker {
     domain: null,
     _events: { message: [Function] },
     _eventsCount: 1,
     _maxListeners: undefined,
     exitedAfterDisconnect: undefined,
     suicide: [Getter/Setter],
     state: 'none',
     id: 1,
     process: 
      ChildProcess {
        domain: null,
        _events: [Object],
        _eventsCount: 5,
        _maxListeners: undefined,
        _closesNeeded: 2,
        _closesGot: 0,
        connected: true,
        signalCode: null,
        exitCode: null,
        killed: false,
        spawnfile: '/Users/akos.kitta/Desktop/electron/theia-electron-app-example/dist/mac/theia-electron-app-example.app/Contents/Frameworks/theia-electron-app-example Helper.app/Contents/MacOS/theia-electron-app-example Helper',
        _handle: [Object],
        spawnargs: [Object],
        pid: 13433,
        stdin: null,
        stdout: null,
        stderr: null,
        stdio: [Object],
        channel: [Object],
        _channel: [Getter/Setter],
        _handleQueue: null,
        _pendingHandle: null,
        send: [Function],
        _send: [Function],
        disconnect: [Function],
        _disconnect: [Function] } } }


inspect
connection undefined
```

The resulting error is the following:
```
module.js:472
    throw err;
    ^

Error: Cannot find module '/Users/akos.kitta/Desktop/electron/theia-electron-app-example/dist/mac/theia-electron-app-example.app/Contents/MacOS/undefined'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:431:7)
    at startup (bootstrap_node.js:155:9)
    at bootstrap_node.js:551:3

```

CC: @akosyakov

--- a/dev-packages/application-package/src/generator/backend-generator.ts
+++ b/dev-packages/application-package/src/generator/backend-generator.ts
@@ -59,9 +59,11 @@ module.exports = (port, host) => Promise.resolve()${this.compileBackendModuleImp
     }
 
     protected compileMain(backendModules: Map<string, string>): string {
-        return ` // @ts-check
+        return `// @ts-check
 const serverPath = require('path').resolve(__dirname, 'server');
-module.exports = require('@theia/core/lib/node/cluster/main').default(serverPath);
+module.exports = require('@theia/core/lib/node/cluster/main').default(serverPath).then(function (address) {
+    process.send(address.port.toString());
+});
 `;
     }
 

--- a/dev-packages/application-package/src/generator/backend-generator.ts
+++ b/dev-packages/application-package/src/generator/backend-generator.ts
@@ -61,9 +61,13 @@ module.exports = (port, host) => Promise.resolve()${this.compileBackendModuleImp
     protected compileMain(backendModules: Map<string, string>): string {
         return `// @ts-check
 const serverPath = require('path').resolve(__dirname, 'server');
-module.exports = require('@theia/core/lib/node/cluster/main').default(serverPath).then(function (address) {
-    process.send(address.port.toString());
+const address = require('@theia/core/lib/node/cluster/main').default(serverPath);
+address.then(function (address) {
+    if (process && process.send) {
+        process.send(address.port.toString());
+    }
 });
+module.exports = address;
 `;
     }
 

--- a/dev-packages/application-package/src/generator/frontend-generator.ts
+++ b/dev-packages/application-package/src/generator/frontend-generator.ts
@@ -87,25 +87,46 @@ if (cluster.isMaster) {
     }
     process.env.LC_NUMERIC = 'C';
     const electron = require('electron');
-    const path = require('path');
     electron.app.on('window-all-closed', function () {
         if (process.platform !== 'darwin') {
             electron.app.quit();
         }
     });
     electron.app.on('ready', function () {
+        const path = require('path');
+        const { fork }  = require('child_process');
+        // Check whether we are in bundled application or development mode.
+        const devMode = process.defaultApp || /node_modules[\\/]electron[\\/]/.test(process.execPath);
         const mainWindow = new electron.BrowserWindow({ width: 1024, height: 728 });
-        require("../backend/main").then(address => {
-            mainWindow.loadURL(\`file://\${path.join(__dirname, '../../lib/index.html')}?port=\${address.port}\`);
-        }).catch(() => {
-            electron.app.exit(1);
-        });
+        const mainPath = path.join(__dirname, '..', 'backend', 'main');
+        const loadMainWindow = function(port) {
+            mainWindow.loadURL(\`file://\${path.join(__dirname, '../../lib/index.html')}?port=\${port}\`);
+        };
+        // We need to distinguish between bundled application and development mode when starting the clusters.
+        // https://github.com/electron/electron/issues/6337#issuecomment-230183287
+        if (devMode) {
+            require(mainPath).then(address => {
+                loadMainWindow(address.port);
+            }).catch((error) => {
+                console.error(error);
+                electron.app.exit(1);
+            });
+        } else {
+            const cp = fork(mainPath, process.argv, { stdio: [0, 1, 2, 'ipc'] });
+            cp.on('message', function (message) {
+                loadMainWindow(message);
+            });
+            cp.on('error', function (error) {
+                console.error(error);
+                electron.app.exit(1);
+            });
+        }
         mainWindow.on('closed', function () {
             electron.app.exit(0);
         });
     });
 } else {
-    require("../backend/main");
+    require('../backend/main');
 }
 `;
     }

